File: apps/api/app/schemas/traceability.py
"""DASH-05 traceability response schemas (mirrors the traceability.chain dict).

BaseModel shapes for GET /api/traceability — the lifecycle chain the viewer renders:
flow ↔ scenario ↔ script ↔ execution ↔ defect, assembled from a single entry artifact id.

Each segment is a LIST (empty = an honest gap the viewer renders as "No {segment} linked.") and the
flow segment is nullable (null = the flow graph is down/absent — never a fabricated node). The
`flow_note` carries the honest degrade reason when the flow segment could not be read. Each segment
item carries its own id + the keys the viewer drills on (run_id/flow_id). Distinct fields from the
dashboards/coverage shapes so the chain shape is never conflated with an aggregate.
"""

from __future__ import annotations

from pydantic import BaseModel


class EntryRef(BaseModel):
    """The picked entry artifact echoed back so the viewer shows what was resolved."""

    type: str | None = None  # flow | scenario | run | defect (None only on an empty echo)
    id: str | None = None


class FlowSegment(BaseModel):
    """The discovered-flow record (best-effort from the graph; the whole segment is nullable)."""

    flow_id: str | None = None
    name: str | None = None
    category: str | None = None
    risk_tier: str | None = None
    step_count: int | None = None


class ScenarioSegment(BaseModel):
    """A generated BDD scenario row in the chain."""

    id: int
    flow_id: str
    run_id: str
    feature_name: str
    status: str


class ScriptSegment(BaseModel):
    """The generated test-project path — CONVENTION-DERIVED from run_id (A4), never a stored column."""

    run_id: str
    path: str
    derived: bool = True


class ExecutionSegment(BaseModel):
    """A per-flow execution result joined to its parent run (tier/status)."""

    run_id: str
    flow_id: str
    verdict: str
    attempts: int
    duration_ms: int | None = None
    tier: str | None = None
    status: str | None = None


class ArtifactSegment(BaseModel):
    """A captured artifact — RUN-RELATIVE path only (the viewer builds the auth-gated URL)."""

    run_id: str
    flow_id: str
    kind: str
    path: str


class DefectSegment(BaseModel):
    """A defect draft/issue in the chain incl. the JIRA-04 jira_key link (nullable until filed)."""

    id: int
    run_id: str
    flow_id: str
    classification: str
    confidence: int
    fingerprint: str
    jira_key: str | None = None
    status: str


class TraceabilityResponse(BaseModel):
    """The assembled lifecycle chain — honest gaps (empty lists / null flow), never fabricated."""

    entry: EntryRef
    flow: FlowSegment | list[FlowSegment] | None = None
    flow_note: str | None = None
    scenarios: list[ScenarioSegment]
    scripts: list[ScriptSegment]
    executions: list[ExecutionSegment]
    artifacts: list[ArtifactSegment]
    defects: list[DefectSegment]

File: apps/web/lib/api/traceability.ts
/**
 * Traceability API surface (plan 10-06 — DASH-05 traceability viewer UI half).
 *
 * Mirrors the backend read-time cross-store join (Plan 03 — services/traceability.py +
 * schemas/traceability.py), role-gated by `require_role` server-side:
 *
 *   GET /api/traceability?{flow_id|scenario_id|run_id|defect_id}=...  -> TraceabilityResponse
 *
 * The viewer picks ONE entry artifact (flow / scenario / run / defect) and the server assembles the
 * lifecycle chain flow ↔ scenario ↔ script ↔ execution ↔ defect on the resolved run_id + flow_id.
 * The router enforces EXACTLY one entry id (422 for zero or multiple); an UNKNOWN id returns an
 * honest empty chain at 200 (the viewer renders the gaps, never a 404). A graph-down degrades the
 * flow segment to null + a `flow_note` (the relational chain still assembles).
 *
 * zod parses every payload at the boundary — every chain node renders STRICTLY from the server; a
 * missing segment is an honest empty list / null gap, never a fabricated node (T-10-30). The api
 * wrapper handles 401 -> refresh -> /login; a 403 the page maps to the no-access state.
 */

import { z } from "zod";

import { api } from "./client";

// --- the entry artifact types the picker offers (map 1:1 to the four query params) -------------

/** The four entry artifact kinds the picker offers (UI label -> the API query param). */
export type EntryType = "flow" | "scenario" | "run" | "defect";

/** Map a picker entry type to its API query-param name (run -> run_id, etc.). */
const ENTRY_PARAM: Record<EntryType, string> = {
  flow: "flow_id",
  scenario: "scenario_id",
  run: "run_id",
  defect: "defect_id",
};

// --- segment schemas (mirror schemas/traceability.py) -----------------------------------------

export const entryRefSchema = z.object({
  type: z.string().nullable().optional(),
  id: z.string().nullable().optional(),
});

export const flowSegmentSchema = z.object({
  flow_id: z.string().nullable().optional(),
  name: z.string().nullable().optional(),
  category: z.string().nullable().optional(),
  risk_tier: z.string().nullable().optional(),
  step_count: z.number().int().nullable().optional(),
});
export type FlowSegment = z.infer<typeof flowSegmentSchema>;

export const scenarioSegmentSchema = z.object({
  id: z.number().int(),
  flow_id: z.string(),
  run_id: z.string(),
  feature_name: z.string(),
  status: z.string(),
});
export type ScenarioSegment = z.infer<typeof scenarioSegmentSchema>;

export const scriptSegmentSchema = z.object({
  run_id: z.string(),
  path: z.string(),
  derived: z.boolean(),
});
export type ScriptSegment = z.infer<typeof scriptSegmentSchema>;

export const executionSegmentSchema = z.object({
  run_id: z.string(),
  flow_id: z.string(),
  verdict: z.string(),
  attempts: z.number().int(),
  duration_ms: z.number().int().nullable().optional(),
  tier: z.string().nullable().optional(),
  status: z.string().nullable().optional(),
});
export type ExecutionSegment = z.infer<typeof executionSegmentSchema>;

export const defectSegmentSchema = z.object({
  id: z.number().int(),
  run_id: z.string(),
  flow_id: z.string(),
  classification: z.string(),
  confidence: z.number().int(),
  fingerprint: z.string(),
  jira_key: z.string().nullable().optional(),
  status: z.string(),
});
export type DefectSegment = z.infer<typeof defectSegmentSchema>;

export const traceabilityResponseSchema = z.object({
  entry: entryRefSchema,
  // The flow segment is a single record, a set (a multi-flow run entry), or null (graph-down/absent).
  flow: z
    .union([flowSegmentSchema, z.array(flowSegmentSchema), z.null()])
    .optional(),
  flow_note: z.string().nullable().optional(),
  scenarios: z.array(scenarioSegmentSchema),
  scripts: z.array(scriptSegmentSchema),
  executions: z.array(executionSegmentSchema),
  artifacts: z.array(
    z.object({
      run_id: z.string(),
      flow_id: z.string(),
      kind: z.string(),
      path: z.string(),
    }),
  ),
  defects: z.array(defectSegmentSchema),
});
export type TraceabilityResponse = z.infer<typeof traceabilityResponseSchema>;

/** Normalize the nullable `flow` union to an array (single record -> [record], null -> []). */
export function flowSegments(resp: TraceabilityResponse): FlowSegment[] {
  if (!resp.flow) return [];
  return Array.isArray(resp.flow) ? resp.flow : [resp.flow];
}

// --- fetcher ----------------------------------------------------------------------------------

/**
 * GET /api/traceability with exactly ONE entry id (role: admin | qa_lead | developer). An unknown
 * id returns an honest empty chain (200); a 422 means the (programming) caller passed != 1 id; a
 * 403 the page maps to no-access.
 */
export async function getTraceability(
  type: EntryType,
  id: string,
): Promise<TraceabilityResponse> {
  const param = ENTRY_PARAM[type];
  const qs = new URLSearchParams({ [param]: id }).toString();
  return traceabilityResponseSchema.parse(await api.get(`/api/traceability?${qs}`));
}
